File: C++ Problems/lengthOfLongestSubstring.cpp
#include<iostream>
#include<unordered_map>

using namespace std;

class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        unordered_map<char, int> stored_vals;
        stored_vals.reserve(s.size());

        int start_index = 0;
        int size = 0;

        for(int i = 0; i < s.size(); i++)
        {
            if(stored_vals.count(s[i]) && stored_vals[s[i]] >= start_index)
            {
                start_index = stored_vals[s[i]] + 1;
            }
            stored_vals[s[i]] = i;
            size = max(size, i - start_index + 1);
        }
        return size;
    }
};

int main()
{
    Solution s;
    cout << s.lengthOfLongestSubstring("abcabcbb");

    return 0;
}

File: Python Problems/lengthOfLongestSubstring.py
class Solution(object):
    def lengthOfLongestSubstring(self, s):
        """
        :type s: str
        :rtype: int
        """
        stored_vals = {}

        size = 0
        start = 0

        for i in range(len(s)):
            if( (s[i] in stored_vals) and (stored_vals[s[i]] >= start) ):
                start = stored_vals[s[i]] + 1

            stored_vals[s[i]] = i
            size = max(size, i - start + 1)
        return size

def main():
    s = "abcadefgh"

    my_obj = Solution()
    print(my_obj.lengthOfLongestSubstring(s))


if (__name__ == "__main__"):
    main()  